projects/feature-activation/0001-feature-activation.md

```diff
@@ -0,0 +1,558 @@
+
+Must be less than or equal to `timeout_height`. Must be an exact multiple of `EVALUATION_INTERVAL` (that is, at an evaluation interval boundary).
+
+#### `activate_on_timeout`
+
+Specifies if the feature should be activated even if the activation criteria are not met when the `timeout_height` is reached, effectively forcing activation. Should be used with caution, only when we are confident about community consensus, and especially confident that miners will comply.
+
+#### `version`
+
+The client version of `hathor-core` at which this feature was defined.
+
+### Examples
+
+Some criteria configuration examples can be found in the table in the [Explorer User Interface] section. There,
+
+- `MY_NEW_FEATURE_1` failed because it didn't reach the threshold before the timeout and `activate_on_timeout` was `False`.
+- `MY_NEW_FEATURE_2` was activated on height 3,200,00 (the `minimum_activation_height`), because it reached the threshold before timeout.
+- `MY_NEW_FEATURE_3`, similarly to `MY_NEW_FEATURE_1`, didn't reach the threshold before the timeout, but in this case it was activated on height 3,300,000 (the `timeout_height`) because `activate_on_timeout` was `True`.
+- `MY_NEW_FEATURE_4` was configured with `threshold = 0`, meaning that it was only dependent on block height, and not on miner support. It was activated at height 3,200,000 (the `minimum_activation_height`).
+- `MY_NEW_FEATURE_5` activation process is ongoing. Its minimum activation height is `0`, meaning that it can be activated as soon as the threshold is reached, if it is before timeout.
+- `MY_NEW_FEATURE_6` activation process is defined but has not started yet. It will start at height 3,500,000.
+
+Note: in those examples, just for illustration purposes, height values are not multiples of `EVALUATION_INTERVAL`. They must be, as defined in the attributes above.
+
+## State transitions
+[State transitions]: #state-transitions
+
+Here's a diagram illustrating how states are updated for each feature and block:
+
+```mermaid
+stateDiagram-v2
+  direction LR
+  DEFINED --> DEFINED
+  DEFINED --> STARTED : height >= start_height
+  STARTED --> STARTED
+  STARTED --> FAILED : height >= timeout_height<br />and not activate_on_timeout
+  STARTED --> ACTIVE : height >= timeout_height<br />and activate_on_timeout<br />and height >= minimum_activation_height
+  STARTED --> ACTIVE : height < timeout_height<br />and threshold reached<br />and height >= minimum_activation_height
+  ACTIVE --> ACTIVE
+  FAILED --> FAILED
+```
+
+## Feature Service
+[Feature Service]: #feature-service
+
+As explained in the [State retrieval] section, a `BaseTransaction` will contain a `feature_service` property of the `FeatureService` class. That class is responsible for handling all logic related to feature activation. There are three methods exposed by the feature activation service:
+
+```python
+class FeatureService:
+    def is_feature_active(self, vertex: BaseTransaction, feature: Feature) -> bool:
```

Comment:
Should it receive a Block instead?

Reply:
It seems likely that we will have features that affect transactions. If I got it right, you will add a helper to transactions so it would get the features from the closest block. In this case, isn't it better to add a method `BaseTransaction.get_feature_service()` and implement something like `return FeatureService(self.get_first_block())`?

Reply:
Indeed it'll be more or less like you described it. But as agreed, feature activation for transactions will be moved to a separate document. For this reason, I changed the `FeatureService` API to only support `Block`s, as you suggested.

Updated in https://github.com/HathorNetwork/rfcs/pull/51/commits/5b4878b3bcfbe2260a607ec8b169f41dab31305e

Reply:
Great! Thanks!